pandas/_libs/src/inference.pyx

```diff
@@ -1319,7 +1319,16 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
 
     # we try to coerce datetime w/tz but must all have the same tz
     if seen.datetimetz_:
-        if len({getattr(val, 'tzinfo', None) for val in objects}) == 1:
```

Comment:
rather than this, I think you can do:

```
if seen.datetimetz_:
    from pandas import to_datetime
    return to_datetime(objects, errors='ignore')
```
and you can remove the ``seen.object_ = 1`` as this will return a DTI if posible or just the input if not (which is what we want)

Reply:
Hi, I tried this, but it didn't pass my parametrized tests, the timezones need to be unique, so this is the proposed fix: 

    # we try to coerce datetime w/tz but must all have the same tz
    if seen.datetimetz_:
        unique_types = set()
        from dateutil import tz
        for val in objects:
            item = getattr(val, 'tzinfo', type(val).__name__)
            # as tzoffset is not hashable, we use __repr__ in our set
            if isinstance(item, tz.tzoffset):
                unique_types.add(item.__repr__())
            else:
                unique_types.add(item)

        if len(unique_types) == 1:
            from pandas import DatetimeIndex
            return DatetimeIndex(objects)
        seen.object_ = 1

Reply:
this needs to change to something more like

```
tzs = (getattr(val, 'tzinfo', None) for val in objects)
if len(set(tzs)) and all(arg for arg in tzs is not None):
    ....
```

you don't need to do all of this checking, we only care if there is 1 and only 1 timezone and no non-timezone aware objects.

Reply:
if I try to implement as you have suggested I get the following compiler error:
```
Error compiling Cython file:
------------------------------------------------------------
...
    return ints


@cython.boundscheck(False)
@cython.wraparound(False)
def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
                         ^
------------------------------------------------------------

pandas/_libs/src/inference.pyx:1196:26: Buffer types only allowed as function local variables
building 'pandas._libs.lib' extension
```
I could get around this by the following: 
```
    tzs = [getattr(val, 'tzinfo', None) for val in objects]
    if len(list(filter(None, tzs))) == 1:
        ...
```

however this does not work as it will fail in the test cases where you have a timezone aware and non-timezone aware dates:
[test.log](https://github.com/pandas-dev/pandas/files/1691820/test.log)